// Copyright 2017 Eryx <evorui аt gmаil dοt cοm>, All rights reserved.

syntax = "proto3";
package lynkstor;


// option optimize_for = LITE_RUNTIME;
option cc_generic_services = true;

// ../deps/protobuf/build/protoc --cpp_out=. lynkstor.proto


message KvMeta {
	uint32 type = 1;
	uint64 version = 2;
	uint64 expired = 3;
	uint64 created = 4;
	uint64 updated = 5;
	uint64 size = 6;
	uint32 sum = 7;
	uint64 num = 8;
	string name = 13;
}

message KvEntry {
	KvMeta meta = 1;
	bytes value = 2;
}

message KvProgKeyEntry {
	uint32 type = 1;
	bytes data = 2;
}

message KvProgKey { repeated KvProgKeyEntry items = 1; }

message KvProgWriteOptions {
	uint64 version = 1;
	uint64 actions = 2;
	uint64 expired = 3;
	uint32 prev_sum = 4;
	uint64 prev_version = 5;
}

message KvProgKeyValueCommit {
	KvMeta meta = 1;
	KvProgKey key = 2;
	bytes value = 3;
	KvProgWriteOptions options = 4;
}

message FileObjectEntryInit {
	string path = 3;
	uint64 size = 4;
	uint64 attrs = 5;
}

message FileObjectEntryMeta {
	KvMeta meta = 1;
	string path = 3;
	uint64 size = 4;
	uint64 attrs = 5;
	uint32 sn = 7;
	string commit_key = 8;
	repeated uint32 blocks = 9;
}

message FileObjectEntryBlock {
	KvMeta meta = 1;
	string path = 3;
	uint64 size = 4;
	uint64 attrs = 5;
	uint32 num = 6;
	uint64 sum = 7;
	uint32 sn = 8;
	bytes data = 9;
	string commit_key = 10;
}

message LynkStorResult {
	int32 status = 1;
	bytes key = 2;
	bytes data = 3;
	int32 cap = 4;
	repeated LynkStorResult items = 5;
	uint64 attrs = 6;
}

message LynkStorRepQuery {
	uint64 data_version = 4;
	string node_id = 5;
	bytes key_offset = 6;
	int32 limit = 9;
}

service LynkStorService {
	rpc SysPing (LynkStorResult) returns (LynkStorResult) {}
	rpc RepFull (LynkStorRepQuery) returns (LynkStorResult) {}
	rpc RepPull (LynkStorRepQuery) returns (LynkStorResult) {}
	rpc KvPut (LynkStorRepQuery) returns (LynkStorResult) {}
	rpc KvGet (LynkStorRepQuery) returns (LynkStorResult) {}
}

message TypeKvPair {
	string key = 1;
	string value = 2;
}

message TypeKvPairs {
	repeated TypeKvPair items = 2;
}

